free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py

```diff
@@ -0,0 +1,390 @@
+from rmf_adapter.robot_update_handle import ActivityIdentifier
+from tf_transformations import euler_from_quaternion, quaternion_from_euler
+
+import zenoh
+
+
+class Nav2TfHandler:
+
+    def __init__(self, robot_name, zenoh_session, tf_buffer, node):
+        self.robot_name = robot_name
+        self.zenoh_session = zenoh_session
+        self.node = node
+        self.tf_buffer = tf_buffer
+
+        def _tf_callback(sample: zenoh.Sample):
+            try:
+                transform = TFMessage.deserialize(sample.payload.to_bytes())
+            except Exception as e:
+                self.node.get_logger().debug(
+                    f'Failed to deserialize TF payload: {type(e)}: {e}'
+                )
+                return None
+            for zt in transform.transforms:
+                t = transform_stamped_to_ros2_msg(zt)
+                t.header.frame_id = namespacify(zt.header.frame_id,
+                                                self.robot_name)
+                t.child_frame_id = namespacify(zt.child_frame_id,
+                                               self.robot_name)
+                self.tf_buffer.set_transform(
+                    t, f'{self.robot_name}_TfListener')
+
+        self.tf_sub = self.zenoh_session.declare_subscriber(
+            namespacify('tf', self.robot_name),
+            _tf_callback
+        )
+
+    def get_transform(self) -> TransformStamped | None:
+        try:
+            # TODO(ac): parameterize the frames for lookup
+            transform = self.tf_buffer.lookup_transform(
+                namespacify('map', self.robot_name),
+                namespacify('base_footprint', self.robot_name),
+                rclpy.time.Time()
+            )
+            return transform
+        except Exception as err:
+            self.node.get_logger().info(
+                'Unable to get transform between base_footprint and map: '
+                f'{type(err)}: {err}'
+            )
+        return None
+
+
+class Nav2RobotAdapter(RobotAdapter):
+
+    def __init__(
+        self,
+        name: str,
+        configuration,
+        robot_config_yaml,
+        node,
+        zenoh_session,
+        fleet_handle,
+        tf_buffer
+    ):
+        RobotAdapter.__init__(self)
+
+        self.name = name
+        self.execution = None
+        self.update_handle = None
+        self.configuration = configuration
+        self.robot_config_yaml = robot_config_yaml
+        self.node = node
+        self.zenoh_session = zenoh_session
+        self.fleet_handle = fleet_handle
+        self.tf_buffer = tf_buffer
+
+        self.nav_goal_id = None
+        self.map = self.robot_config_yaml['initial_map']
+
+        # TODO(ac): Only use full battery if sim is indicated
+        self.battery_soc = 1.0
+
+        self.replan_counts = 0
+
+        self.tf_handler = Nav2TfHandler(
+            self.name,
+            self.zenoh_session,
+            self.tf_buffer,
+            self.node
+        )
```

Comment:
I assume the intention of using an abstract class `RobotAdapter` is to open up options for integrators who want to tweak the implementation of the Nav2 adapter according to their use case, e.g. adding custom actions in the future.
In that case, some of these parameters can be ported into `RobotAdapter` to ensure that they will be initialized instead of relying on integrators to set them and possibly missing them out. It could also be useful to document this intention inside README to direct integrators to use this abstract class.

This also looks good to brought into the [fleet_adapter_template](https://github.com/open-rmf/fleet_adapter_template/blob/main/fleet_adapter_template/fleet_adapter_template/RobotClientAPI.py) with some tweaks, but that's a goal for another day.

Reply:
I would actually leave the setting of member variables to the child classes, in this case `Nav2RobotAdapter`, instead of moving these operations upward to `RobotAdapter`. Keeping `RobotAdapter` as a pure abstract class will help us not limit the constructors of any implemented `RobotAdapter`, but still maintain an overridden API that the final free fleet adapter can call upon

I believe this will help us when we start implementing the `Nav1RobotAdapter`. But in any case, if I don't see any major benefit down the road, I'll follow your advice and help move these upwards into `RobotAdapter` and update `fleet_adapter_template` too

Reply:
Gotcha, I understand where you're coming from. Flexibility is definitely a desirable trait we want in these classes. Regarding member variables, params such as `update_handle`, `fleet_handle` and `node` (and maybe even robot `name`) are pretty  much non-negotiable for any EFC fleet adapter. Having them defined in the abstract class would ensure that users provide them as arguments when setting up their fleet adapters. The alternate case could be users sifting through existing fleet adapter implementations to pick out variables they assume are important, and potentially leaving out these "essentials", making the setup process longer and more confusing. I'm not leaning very strongly to have these added in though, just a suggestion.

If you're reconsidering including them but are still worried about constructor limitations, I recently received a [review comment](https://github.com/open-rmf/fleet_adapter_mir/pull/36#discussion_r1784333490) that suggested wrapping variables in an [`ActionContext`](https://github.com/open-rmf/fleet_adapter_mir/blob/main/fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py#L71) object, and [import it into the abstract class](https://github.com/open-rmf/fleet_adapter_mir/blob/main/fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py#L21-L22) as a single argument. I thought the concept could be useful for this scenario too, allowing us to update the Context object without breaking API if we ever want to add on to the non-nego variables.

With that said, I still struggle to see the need for a pure abstract `RobotAdapter` class that only lives within `free_fleet` when it has nothing specific to free fleet, and instead can be utilized across all types of EFC fleet adapters for different robots. Imo it would be more suitable to port this to the fleet adapter template, and have free fleet adapters inherit the class from there. That way we won't have multiple abstract class across all our fleet adapter repos. If we do want an abstract `RobotAdapter` class under the `free_fleet` repo, it should probably contain some `free_fleet`-specific methods.

Reply:
Ok, there are a few dilemmas here, because the `RobotAdapter` is for testability and consistency over the nav1 and nav2 robot adapters (and any other types of navigation stack we might consider in the future), and not a plugin system,

> potentially leaving out these "essentials"

the essentials can technically be left out if their fleet adapter implementations don't require it (I can't think of why, but it is possible to leave out name and node, and just have the fleet adapter keep track of it)

users might also want to create their own node inside their own `RobotAdapter` perhaps to set up different callback groups for their own uses or optimizations, instead of using the provided node like we usually do

> If you're reconsidering including them but are still worried about constructor limitations

Since the next few items on the list for `free_fleet` is related to implementing the nav1 robot adapter, what exactly can be abstracted into the constructor `RobotAdapter` is yet to be confirmed, and will have a high risk of changing, breaking API. I will look into using `ActionContext`, but I'd say we can leave that for later

> the need for a pure abstract RobotAdapter class that only lives within free_fleet when it has nothing specific to free fleet

The `Nav1RobotAdapter` will be implemented in the same manner such that in the future, the `free_fleet_adapter` is able to control a mixed fleet of Nav1 and Nav2 robots, so I'd say it is rather related to free fleet, although as you said, not specific only to free fleet.

On the other hand `fleet_adapter_template` at the moment is a template, and the final implementation of any fleet adapter based on the template still needs to be written in a separate package, without importing it from `fleet_adapter_template`. Adding an abstracted `RobotAdapter` to `fleet_adapter_template`, although useful as a reference, could lead to even more confusion, as folks can import `RobotAdapter` but implement their own fleet adapter based on the `fleet_adapter_template` at that time . But if any implementation details (member variable name) changes upstream for the imported `RobotAdapter`, without updating their own fleet adapter, things will break. Hence I think it will be clearer to users if they have full visibility of `RobotAdapter`.

I've moved the member variables directly used by the `free_fleet_adapter` into `RobotAdapter`, while keeping everything else API calls (which have been abstracted). I'd say only if we can make the upstream fleet adapter template a library or a plugin system for robot adapters, should we consider moving `RobotAdapter` upstream